jq255.md

```diff
@@ -0,0 +1,1405 @@
+# The jq255e and jq255s Groups, Key Exchange and Signatures
+
+  - Version: 0.0.0
+  - Author: Thomas Pornin
+
+## Introduction
+
+This document specifies the **jq255e** and **jq255s** groups. These are
+prime order groups based on elliptic curves, with characteristics
+suitable for building cryptographic protocols:
+
+  - The order of each group is a prime integer close to `2^254`, hence
+    formally offering "127-bit security" against discrete logarithm
+    attacks.
+
+  - Each group element (including the neutral element) can be encoded as
+    a 32-byte sequence. The encoding is canonical: a given element can
+    yield only a single possible sequence of 32 bytes, and the decoding
+    process verifies that the encoding was done properly. Decoding
+    cannot yield an out-of-group element, so that invalid curve attacks
+    are not possible, and there are no cofactor issues.
+
+  - Operations on group elements can be implemented with complete
+    formulas that have no exceptional case, and are amenable to
+    constant-time implementations.
+
+  - Operations are efficient, both on large powerful CPUs and on small
+    microcontrollers. The jq255e group is furthermore defined over a
+    curve that admits an easily computed non-trivial endomorphism, that
+    can be leveraged to further speed up multiplication of group
+    elements by scalars.
+
+On top of these groups, the following primitives are specified by this
+document:
+
+  - A digital signature scheme. Signatures have size 48 bytes, which is
+    shorter than the more usual 64-byte signatures associated with
+    elliptic curves of similar security (e.g. EdDSA signatures on
+    Curve25519, or ECDSA over P-256 or secp256k1 curves), but still
+    provides the security level expected from the curve. Signature
+    verification is also faster than with the common 64-byte signatures.
+
```

Comment:
The half-size challenge signature scheme defined below is interesting - would it be useful to break it out into its own group and hash agnostic scheme, as long as the group and hash properties are constrained?

Reply:
Technically you can define such a scheme over any prime-order group (that's basically what Schnorr did in the first place!) and you could do that over, for instance, Ristretto. But here I am trying to have something self-contained that does not spill over several spec documents, since such dispersion seems to lead only to the spec being ignored (conversely, I always found that the great advantage of RFC 8032 is that it did everything in a single document). Also, there are some shortcuts that I can take here that are interesting for implementations, especially on small microcontrollers, namely that I can take a 32-byte random blob and reduce it modulo r, and it's unbiased enough; I cannot do that generically, but for jq255e and jq255s it works because r is very close to a power of 2. A generic definition for all groups would have to "play safe" and use a 48-byte output and that would mean BLAKE2b or a BLAKE2s-based XOF with several invocations, and more generic modular reduction, and that would be inconvenient for implementers (and increase code footprint).

Reply:
@str4d @FiloSottile thoughts?

Reply:
On the one hand, I'd like C2SP to define reusable components across clean abstractions like prime-order groups. On the other hand, I really dislike the "least common multiple" effect of having to bring in the complexity required to handle all possible concrete implementations. It's a tension. I feel like a way to optimize that is not aiming for runtime composition (aka cipher suites) but "design-time" composition.

This specific shortcut though feels like it can be addressed as part of the abstraction: you define a map from a fixed number of uniform bytes to a uniform scalar, and for these groups the fixed number of bytes is 32.

I tend to agree having everything in a document is advantageous, but maybe that can be achieved by reference, like "signatures over these groups are constructed by applying style X of the prime-order group signatures document with these parameters" followed by test vectors that lock them in.

Reply:
I have added a formal definition of `UNIFORM_SIZE(r)`, which is the number of random bytes you need so that the reduce-modulo-r process yields something which is indistinguishable from uniform selection. I have included some formulas that should be fine for every possible r, and make `UNIFORM_SIZE(r) = 32` for both jq255e and jq255s. Then I added references to that function where appropriate (private key generation, and signature nonce generation). Thus, I get the optimization for my case, but I still document the general case so that extension to larger curves is possible.

More generally, I think that _in the long term_ there should be a generic "sign over a prime-order group" specification that works for everything (with some interactions to unravel with both RFC 8032 and the FROST draft spec); when such a document exist, the jq255 spec can be modified to reference that document. In the short term, though, the jq255 spec may include the high-level algorithms (as it does presently), if only to be able to gather experience and comments from outsiders. Isn't it why C2SP specs are versioned, after all?